ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+    return getCurrentCycleLog().compactionDetails;
+  }
+
+  /**
+   * @return the {@link StoreFindToken} until which data has been copied and flushed.
+   */
+  StoreFindToken getSafeToken() {
+    return getCurrentCycleLog().safeToken;
+  }
+
+  /**
+   * Sets the {@link StoreFindToken} until which data is copied and flushed.
+   * @param safeToken the {@link StoreFindToken} until which data is copied and flushed.
+   */
+  void setSafeToken(StoreFindToken safeToken) {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getState().equals(State.COPY)) {
+      throw new IllegalStateException("Cannot set a safe token - not in COPY phase");
+    }
+    cycleLog.safeToken = safeToken;
+    flush();
+  }
+
+  /**
+   * Marks the start of the copy phase.
+   */
+  void markCopyStart() {
+    getCurrentCycleLog().copyStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the switch phase.
+   */
+  void markSwitchStart() {
+    getCurrentCycleLog().switchStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the cleanup phase.
+   */
+  void markCleanupStart() {
+    getCurrentCycleLog().cleanupStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the current compaction cycle as complete.
+   */
+  void markCycleComplete() {
+    getCurrentCycleLog().cycleEndTime = time.milliseconds();
+    currentIdx++;
+    flush();
+  }
+
+  /**
+   * Closes the compaction log. If compaction is complete, renames the log to keep a permanent record.
+   */
+  @Override
+  public void close() {
+    if (file.exists() && getCompactionState().equals(State.DONE)) {
+      File savedLog = new File(file.getAbsolutePath() + "_" + startTime);
+      if (!file.renameTo(savedLog)) {
+        throw new IllegalStateException("Compaction log could not be renamed after completion of compaction");
+      }
+    }
+  }
+
+  /**
+   * @return the {@link CycleLog} for the current compaction cycle.
+   */
+  private CycleLog getCurrentCycleLog() {
+    return cycleLogs.get(currentIdx);
+  }
+
+  /**
+   * Flushes all changes to the file backing this compaction log.
+   */
+  private void flush() {
+    /*
+        Description of serialized format
+
+        version
+        startTime
+        index of current cycle's log
+        size of cycle log list
+        cycleLog1 (see CycleLog#toBytes())
+        cycleLog2
+        ...
+        crc
+       */
+    File tempFile = new File(file.getAbsolutePath() + ".tmp");
+    try (FileOutputStream fileOutputStream = new FileOutputStream(tempFile)) {
+      CrcOutputStream crcOutputStream = new CrcOutputStream(fileOutputStream);
+      DataOutputStream stream = new DataOutputStream(crcOutputStream);
+      stream.writeShort(VERSION_0);
+      stream.writeLong(startTime);
+      stream.writeInt(currentIdx);
+      stream.writeInt(cycleLogs.size());
+      for (CycleLog cycleLog : cycleLogs) {
+        stream.write(cycleLog.toBytes());
```

Comment:
Help me if I am missing something. I see that we flush every detail of every cycleLog(which might require renaming) to disk. But why do we need every detail about a completed cycle? For eg, if one compaction has 3 cycles in it, once we move to cycle2, do we need any details from the previous cycle at any point in time? Atleast according to this class, it looks like the caller will always be interested in current cycleLog's details only and will update the same(setSafeToken, markCopyStart, markSwitchStart and markCleanupStart). If possible, we should cut down on those unnecessary details which may not be required at all.